ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/QueryResultImpl.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.api.client.impl;
+
+import io.confluent.ksql.api.client.QueryResult;
+import io.confluent.ksql.api.client.Row;
+import io.confluent.ksql.reactive.BufferedPublisher;
+import io.vertx.core.Context;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.reactivestreams.Subscriber;
+
+class QueryResultImpl extends BufferedPublisher<Row> implements QueryResult {
+
+  private final String queryId;
+  private final List<String> columnNames;
+  private final List<String> columnTypes;
+  private final PollableSubscriber pollableSubscriber;
+  private boolean polling;
+  private boolean subscribing;
+
+  QueryResultImpl(final Context context, final String queryId, final List<String> columnNames,
+      final List<String> columnTypes) {
+    super(context);
+    this.queryId = queryId;
+    this.columnNames = columnNames;
+    this.columnTypes = columnTypes;
+    this.pollableSubscriber = new PollableSubscriber(ctx);
+  }
+
+  @Override
+  public List<String> columnNames() {
+    return columnNames;
+  }
+
+  @Override
+  public List<String> columnTypes() {
+    return columnTypes;
+  }
+
+  @Override
+  public String queryID() {
+    return queryId;
+  }
+
+  @Override
+  public synchronized void subscribe(final Subscriber<? super Row> subscriber) {
+    if (polling) {
+      throw new IllegalStateException("Cannot set subscriber if polling");
+    }
+    subscribing = true;
```

Comment:
Probably be better to take the monitor after the check if polling, otherwise if polling, the method won't be executed and the user won't see the exception.

Reply:
Sorry, I'm failing to grok this comment. What does "the monitor" refer to?

Reply:
Each Java object which uses synchronized somewhere has a monitor associated with it. You can think of it as a mutex which can be owned by a thread and must be released by the same thread. It's what does the locking when you use the synchronized keyword.
So when I say "take the monitor" I mean "lock the object". In other words instead of synchronizing the whole method, do the polling check outside the monitor, something like this:
```
  @Override
  public void subscribe(final Subscriber<? super Row> subscriber) {
    if (polling) {
      throw new IllegalStateException("Cannot set subscriber if polling");
    }
    synchronized (this) {
        subscribing = true;
        super.subscribe(subscriber);
   }
  }
```
polling would need to made volatile too.
Sorry about the confusing language!

Reply:
Got it -- good catch! Made the suggested change.